lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp

```diff
@@ -0,0 +1,476 @@
+      break;
+    case MemOp::PortKind::ReadWrite:
+      prefix = (Twine("RW") + Twine(readWritePorts++) + "_").str();
+      break;
+    }
+    // Flatten the bundle representing a memory port, name-mangling and adding
+    // every field in the bundle to the exter module's port list.
+    auto type = op.getResult(i).getType().cast<FIRRTLType>();
+    if (type.isa<FlipType>())
+      type = type.cast<FlipType>().getElementType();
+    for (auto bundleElement : type.cast<BundleType>().getElements()) {
+      auto name = builder.getStringAttr(prefix + bundleElement.name.str());
+      auto type = FlipType::get(bundleElement.type);
+      extPorts.push_back({name, type});
+    }
+  }
+}
+
+/// Create an external module blackbox representing the memory operation.
+/// Returns the port list of the external module.
+static FExtModuleOp
+createBlackboxModuleForMem(MemOp op,
+                           const SmallVectorImpl<ModulePortInfo> &extPorts) {
+
+  OpBuilder builder(op->getContext());
+
+  // The module's name is the name of the memory postfixed with "_ext".
+  StringRef memName = "mem";
+  if (op.name().hasValue())
+    memName = op.name().getValue();
+  std::string extName = memName.str() + "_ext";
+
+  // Create the blackbox external module.
+  auto extModuleOp = builder.create<FExtModuleOp>(
+      op.getLoc(), builder.getStringAttr(extName), extPorts);
+
+  // Insert the external module into the circuit.  This will rename the
+  // external module if there is a conflict with another module name.
+  auto circuitOp = op->getParentOfType<CircuitOp>();
+  SymbolTable symbolTable(circuitOp);
+  symbolTable.insert(extModuleOp);
+
+  // Move the external module to the beginning of the circuitOp.  Inserting into
+  // the symbol table may have moved the operation to the end of the circuit.
+  extModuleOp->moveBefore(&circuitOp.front());
+
+  // Add the memory parameters as attributes on the external module.  These will
+  // be used by the generator tools to create memory configuration files, or
+  // create the memory itself.  We use a named attr list to avoid creating lots
+  // of intermediate dictionary attributes.
+  NamedAttrList attrs(extModuleOp->getAttrDictionary());
+  attrs.set("generator", builder.getStringAttr("FIRRTLMemory"));
+  attrs.set("readLatency", op.readLatencyAttr());
+  attrs.set("writeLatency", op.writeLatencyAttr());
+  attrs.set("depth", op.depthAttr());
+  attrs.set("ruw", op.ruwAttr());
+  extModuleOp->setAttrs(attrs.getDictionary(builder.getContext()));
+
+  return extModuleOp;
+}
+
+/// Create a regular module to wrap the external module.  The wrapper module
+/// instantiates the external module, and connects all the inputs and outputs
+/// together. This module will match the bundle return type of the memory op,
+/// and connects to the flattened parameters of the external module. This is
+/// done for compatibility with the SFC and it is unclear if this will be needed
```

Comment:
nit: Expand SFC, that's not a standard term

Reply:
fixed, thanks!